### PR TITLE
Provider transformed to use gRPC client

### PR DESCRIPTION
**Pull request by jdziurzynski**
KNKT-257
Terraform Cloud Export provider transformed to use new gRPC client.
For test purpose implemented fake_server (provider/fake_server_test.go). Each test running own fake_server instance.
Due to the fake_server specification, the examples can only be run in production mode. The run.sh scripts in the examples have been removed as they are no longer needed when only one mode is available.

### CODE REVIEW

**Review by mateuszmidor — APPROVED** [review]

**Review by danielfurman — CHANGES_REQUESTED** [review]

**Comment by danielfurman** [comment]
> Due to the fake_server specification, the examples can only be run in production mode.

I don't agree with that. We could run examples with fake_server, but it would require writing `main` function for it. It might be done in the future if we need that, so I suggest to leave run.sh scripts intact.

File: README.md [hunk 1]
```diff
@@ -28,18 +28,13 @@ Development steps:
 
 ### Test
 
-Tests run the provider against a `test-api-server` that serves data read from [test-data.json](internal/provider/test-data.json)
+Tests run the provider against a `fakeCloudExportServer`
```
Comment (mmac-m3a) [inline README.md:31]: I like the name `test-api-server` much better than `fakeCloudExportServer`.

File: internal/provider/cloudexport_schema.go [hunk 2]
```diff
@@ -399,48 +383,63 @@ func resourceDataToCloudExport(d *schema.ResourceData) (*cloudexport.V202101beta
...
+			}
+			export.Properties = &ibm
 		}
 	default:
-		return nil, fmt.Errorf("cloud_provider should be one of [aws, azure, ibm, gce, bgp], got: %q", cloudProvider)
+		return nil, fmt.Errorf("cloud_provider should be one of [aws, azure, ibm, gce], got: %q", cloudProvider)
 	}
-
-	return export, nil
+	if cloudProvider == "bgp" {
```
[the remark this answers is not in the document]
Comment (mmac-m3a) [reply internal/provider/cloudexport_schema.go:433]: @danielfurman, you are right. BGP property is separate from provider type.

File: internal/provider/fake_server_test.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+func newFakeCloudExportServer(t testing.TB, ces []*cloudexportpb.CloudExport) *fakeCloudExportServer {
+	return &fakeCloudExportServer{
+		done: make(chan struct{}),
+		t:    t,
+		data: ces,
+	}
+}
+
+func (s *fakeCloudExportServer) Start() {
+	l, err := net.Listen("tcp", "localhost:0")
```
Comment (mmac-m3a) [inline internal/provider/fake_server_test.go:39]: Is 0 really the correct port to listen on?

File: internal/provider/fake_server_test.go [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+	return &cloudexportpb.DeleteCloudExportResponse{}, nil
+}
+
+func (s *fakeCloudExportServer) allocateNewID() string {
+	var id int
+
+	for _, item := range s.data {
+		itemID, err := strconv.Atoi(item.Id)
+		if err != nil {
+			itemID = 1000000 // str conversion error, assume some high integer for the id
```
Comment (mmac-m3a) [inline internal/provider/fake_server_test.go:132]: Why is this needed? Can it even work? I'd prefer to print a descriptive error message and terminate if the `item.Id` string is not a valid integer.

File: internal/provider/provider.go [hunk 5]
```diff
@@ -171,10 +175,9 @@ func getRetryConfig(ctx context.Context, d *schema.ResourceData) (kentikapi.Retr
 	// KentikAPI returns http 500 when creating CloudExport with name that is already taken.
 	// This is non-retryable situation, so exclude http 500 from RetryableStatusCodes
 	return kentikapi.RetryConfig{
-		MaxAttempts:          &maxAttempts,
-		MinDelay:             &minDelay,
-		MaxDelay:             &maxDelay,
-		RetryableStatusCodes: []int{429, 502, 503, 504},
+		MaxAttempts: &maxAttempts,
+		MinDelay:    &minDelay,
+		MaxDelay:    &maxDelay,
```
Comment (mmac-m3a) [inline internal/provider/provider.go:180]: Does it mean that every unsuccessful operation will be retried regardless of the type of error? AFAIk, gRPC all has response error codes.

File: internal/provider/provider_test.go [hunk 6]
```diff
@@ -72,9 +97,17 @@ const (
...
 			id = "1"
 		}
-	`
-	invalidRetryConfig = `
+	`,
+		apiURL,
+	)
+}
+
+func makeInvalidRetryConfig(apiURL string) string {
```
Comment (mmac-m3a) [inline internal/provider/provider_test.go:105]: What is "invalid" about the retry config bellow?

File: internal/provider/fake_server_test.go (hunk 3 above)
Comment (jdziurzynski) [reply internal/provider/fake_server_test.go]: When port is set to 0, system run server on random suitable port number.

File: internal/provider/provider_test.go (hunk 6 above)
Comment (jdziurzynski) [reply internal/provider/provider_test.go:105]: max_delay is invalid, set to: max_delay = "invalid-delay"

Comment (mmac-m3a) [reply internal/provider/provider_test.go:105]: I see.

File: internal/provider/fake_server_test.go (hunk 3 above)
Comment (mmac-m3a) [reply internal/provider/fake_server_test.go]: How does the API client used in test then know to which port to connect?

File: internal/provider/provider.go (hunk 5 above)
Comment (jdziurzynski) [reply internal/provider/provider.go:180]: makeRetryInterceptor() determines on which error code make retries. Only in this interceptor we can configure it.
https://github.com/kentik/community_sdk_golang/blob/2fb7ec871b8d1e6bdcde6f5e86defb49dda40237/kentikapi/client.go#L268

File: internal/provider/cloudexport_schema.go [hunk 7]
```diff
@@ -428,18 +435,18 @@ func resourceDataToCloudExport(d *schema.ResourceData) (*models.CloudExport, err
...
-		}
-		applyBGP := providerMap["apply_bgp"]
-		if v, ok := applyBGP.(bool); ok {
-			bgp.ApplyBGP = &v
-		}
-		export.BGP = &bgp
+	return nil
+}
+
+func resourceDataToBGPProperties(export *models.CloudExport, m map[string]interface{}) {
```
Comment (danielfurman) [inline internal/provider/cloudexport_schema.go:443]:
```suggestion
func resourceDataToBGPProperties(data interface{}) (<bgp>, error) {}
```
The same as above: please modify the function to return bgp. Example conversion function for nested object: https://github.com/kentik/terraform-provider-kentik-synthetics/blob/c9ba2f17883e7a3e04cc504a18a85d7d122f5f9c/synthetics/test_conversions.go#L511
Function parameters might look a little bit different, of course.

File: internal/provider/cloudexport_schema.go [hunk 8]
```diff
@@ -328,30 +328,25 @@ func cloudExportToMap(e *models.CloudExport) map[string]interface{} {
 }
 
 // resourceDataToCloudExport is used for API create/update operations to fill cloudexport item from terraform resource.
-//nolint: gocyclo
 func resourceDataToCloudExport(d *schema.ResourceData) (*models.CloudExport, error) {
 	// Note: only set the user-writable attributes, read-only attributes that are only generated on server side:
```
Comment (danielfurman) [inline internal/provider/cloudexport_schema.go:332]:
```suggestion
	// Note: only set the user-writable attributes and ID. Read-only attributes that are only generated on server side:
```

File: internal/provider/resource_cloudexport.go [hunk 9]
```diff
@@ -88,6 +90,6 @@ func resourceCloudExportDelete(ctx context.Context, d *schema.ResourceData, m in
 	if err != nil {
 		return detailedDiagError("Failed to delete cloud export", err)
 	}
-	tflog.Debug(ctx, "Deleted cloud export Kentik API response")
+	tflog.Debug(ctx, fmt.Sprintf("Deleted cloud export with ID %v in Kentik", d.Get("id").(string)))
```
Comment (danielfurman) [inline internal/provider/resource_cloudexport.go:93]: Please use the API of Debug() function fully, just like in other places (e.g. L53), instead of using Sprintf.

**Review by danielfurman — CHANGES_REQUESTED** [review]

File: Makefile [hunk 10]
```diff
@@ -32,17 +31,6 @@ lint:
...
-	# - set KTAPI_URL to our test API server URL - otherwise the provider will try to connect to live Kentik server
-	# - set KTAPI_AUTH_EMAIL and KTAPI_AUTH_TOKEN to dummy values - they are required by provider,
-	#   but not actually used by test API server
-	# - set no test caching (-count=1) - beside the provider itself, the tests also depend on the
-	#   test API server and test data
-	KTAPI_URL="http://${TEST_API_SERVER_ENDPOINT}" KTAPI_AUTH_EMAIL="dummy@acme.com" KTAPI_AUTH_TOKEN="dummy" \
-		go test ./... $(TESTARGS) -timeout=5m -count=1 || (./tools/stop_test_api_server.sh && exit 1)
-
-	 ./tools/stop_test_api_server.sh
+	go test ./... -timeout=5m -count=1
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply Makefile]: Done

File: tools/utility_functions.sh [hunk 11]
```diff
@@ -49,46 +49,3 @@ function colored_echo() {
         echo "$msg"
     fi
 }
-
-function run_examples() {
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply tools/utility_functions.sh]: Done

File: tools/tools.go [hunk 12]
```diff
@@ -4,7 +4,7 @@ package tools
 
 //nolint:typecheck
 import (
-	_ "github.com/hashicorp/terraform-plugin-docs/cmd/tfplugindocs"    // documentation generation
-	_ "github.com/kentik/community_sdk_golang/kentikapi/fakeapiserver" // used for tests
-	_ "mvdan.cc/gofumpt"                                               // code formatting
+	_ "github.com/hashicorp/terraform-plugin-docs/cmd/tfplugindocs" // documentation generation
+	//_ "github.com/kentik/community_sdk_golang/kentikapi/fakeapiserver" // used for tests
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply tools/tools.go]: Done

File: internal/provider/cloudexport_schema.go [hunk 13]
```diff
@@ -347,46 +328,49 @@ func cloudExportToMap(e *cloudexport.V202101beta1CloudExport) map[string]interfa
...
+	if v, ok := planID.(string); ok {
+		export.PlanID = v
+	}
 
 	// required
 	cloudProvider := d.Get("cloud_provider").(string) //nolint: errcheck, forcetypeassert // type enforced by schema
-	export.SetCloudProvider(cloudProvider)
+	export.CloudProvider = models.CloudProvider(cloudProvider)
 
 	// validation: for any given cloud_provider, there should also be an object of the same name,
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply internal/provider/cloudexport_schema.go:385]: Partially done. I created follow-up comment.

File: internal/provider/cloudexport_schema.go [hunk 14]
```diff
@@ -399,48 +383,63 @@ func resourceDataToCloudExport(d *schema.ResourceData) (*cloudexport.V202101beta
...
+			aws := models.AWSProperties{
+				Bucket:     providerMap["bucket"].(string),
+				IAMRoleARN: providerMap["iam_role_arn"].(string),
+				Region:     providerMap["region"].(string),
+			}
+			deleteAfterRead := providerMap["delete_after_read"]
+			if v, ok := deleteAfterRead.(bool); ok {
+				aws.DeleteAfterRead = &v
+			}
+			multipleBuckets := providerMap["multiple_buckets"]
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply internal/provider/cloudexport_schema.go]: Done

File: internal/provider/cloudexport_schema.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (danielfurman) [reply internal/provider/cloudexport_schema.go]: Done

File: internal/provider/data_source_cloudexport_item.go [hunk 15]
```diff
@@ -19,22 +19,19 @@ func dataSourceCloudExportItem() *schema.Resource {
 
 func dataSourceCloudExportItemRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
 	tflog.Debug(ctx, "Get cloud export Kentik API request", map[string]interface{}{"ID": d.Get("id").(string)})
-	getResp, httpResp, err := m.(*kentikapi.Client).CloudExportAdminServiceAPI.
-		ExportGet(ctx, d.Get("id").(string)).
-		Execute()
+	getResp, err := m.(*kentikapi.Client).CloudExports.Get(ctx, d.Get("id").(string))
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply internal/provider/data_source_cloudexport_item.go]: Done

File: internal/provider/fake_server_test.go [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+	}, nil
+}
+
+func (s *fakeCloudExportServer) GetCloudExport(
+	ctx context.Context, req *cloudexportpb.GetCloudExportRequest,
+) (*cloudexportpb.GetCloudExportResponse, error) {
+	if idx := s.findByID(req.GetId()); idx != cloudExportNotFound {
+		return &cloudexportpb.GetCloudExportResponse{Export: s.data[idx]}, nil
+	}
+	return nil, fmt.Errorf("cloud export with ID %q not found", req.GetId())
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply internal/provider/fake_server_test.go]: Done

File: internal/provider/fake_server_test.go [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+func (s *fakeCloudExportServer) CreateCloudExport(
+	ctx context.Context, req *cloudexportpb.CreateCloudExportRequest,
+) (*cloudexportpb.CreateCloudExportResponse, error) {
+	newExport := req.GetExport()
+
+	if s.findByName(newExport.Name) != cloudExportNotFound {
+		return nil, fmt.Errorf("cloud export %q already exists", newExport.Name)
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply internal/provider/fake_server_test.go]: Done

File: internal/provider/fake_server_test.go [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+	ctx context.Context, req *cloudexportpb.UpdateCloudExportRequest,
+) (*cloudexportpb.UpdateCloudExportResponse, error) {
+	exportUpdate := req.GetExport()
+	if i := s.findByID(exportUpdate.GetId()); i != cloudExportNotFound {
+		s.data[i] = exportUpdate
+		return &cloudexportpb.UpdateCloudExportResponse{
+			Export: exportUpdate,
+		}, nil
+	}
+	return nil, fmt.Errorf("cloud export of id %q doesn't exists", exportUpdate.Id)
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply internal/provider/fake_server_test.go]: Done

File: internal/provider/fake_server_test.go [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+func (s *fakeCloudExportServer) DeleteCloudExport(
+	ctx context.Context, req *cloudexportpb.DeleteCloudExportRequest,
+) (*cloudexportpb.DeleteCloudExportResponse, error) {
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply internal/provider/fake_server_test.go]: Done

File: internal/provider/resource_cloudexport.go [hunk 20]
```diff
@@ -28,43 +26,34 @@ func resourceCloudExportCreate(ctx context.Context, d *schema.ResourceData, m in
...
 
 func resourceCloudExportRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
 	tflog.Debug(ctx, "Get cloud export Kentik API request", map[string]interface{}{"ID": d.Get("id").(string)})
-	resp, httpResp, err := m.(*kentikapi.Client).CloudExportAdminServiceAPI.
-		ExportGet(ctx, d.Get("id").(string)).
-		Execute()
+	resp, err := m.(*kentikapi.Client).CloudExports.Get(ctx, d.Get("id").(string))
+
 	tflog.Debug(ctx, "Get cloud export Kentik API response", map[string]interface{}{"response": resp})
 	if err != nil {
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply internal/provider/resource_cloudexport.go:54]: After the modification, the resource is deleted in TF state on all API errors. It should be removed only on NotFound error, just like it was before! Also NotFound is not an error case. Please just keep the logic as it was, but change only details of error check.

File: internal/provider/resource_cloudexport.go [hunk 21]
```diff
@@ -28,43 +26,34 @@ func resourceCloudExportCreate(ctx context.Context, d *schema.ResourceData, m in
...
 	// read back the just-created resource to handle the case when server applies modifications to provided data
 	return resourceCloudExportRead(ctx, d, m)
 }
 
 func resourceCloudExportRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
 	tflog.Debug(ctx, "Get cloud export Kentik API request", map[string]interface{}{"ID": d.Get("id").(string)})
-	resp, httpResp, err := m.(*kentikapi.Client).CloudExportAdminServiceAPI.
-		ExportGet(ctx, d.Get("id").(string)).
-		Execute()
+	resp, err := m.(*kentikapi.Client).CloudExports.Get(ctx, d.Get("id").(string))
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply internal/provider/resource_cloudexport.go]: Done

File: internal/provider/resource_cloudexport.go [hunk 22]
```diff
@@ -28,43 +26,34 @@ func resourceCloudExportCreate(ctx context.Context, d *schema.ResourceData, m in
...
-	req := *cloudexport.NewV202101beta1CreateCloudExportRequest()
-	req.Export = export
-	tflog.Debug(ctx, "Create cloud export Kentik API request", map[string]interface{}{"request": req})
-	resp, httpResp, err := m.(*kentikapi.Client).CloudExportAdminServiceAPI.
-		ExportCreate(ctx).
-		Body(req).
-		Execute()
+	tflog.Debug(ctx, "Create cloud export Kentik API request", map[string]interface{}{"request": export})
+
+	resp, err := m.(*kentikapi.Client).CloudExports.Create(ctx, export)
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply internal/provider/resource_cloudexport.go]: Done

File: internal/provider/resource_cloudexport.go [hunk 23]
```diff
@@ -101,13 +84,10 @@ func resourceCloudExportUpdate(ctx context.Context, d *schema.ResourceData, m in
...
-		ExportDelete(ctx, d.Get("id").(string)).
-		Execute()
-	tflog.Debug(ctx, "Delete cloud export Kentik API response", map[string]interface{}{"response": httpResp.Body})
+	err := m.(*kentikapi.Client).CloudExports.Delete(ctx, d.Get("id").(string))
 	if err != nil {
-		return detailedDiagError("Failed to delete cloud export", err, httpResp)
+		return detailedDiagError("Failed to delete cloud export", err)
 	}
-
+	tflog.Debug(ctx, "Deleted cloud export Kentik API response")
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply internal/provider/resource_cloudexport.go]: Done. Added a follow-up comment, because actually tflog encourages structured logging.

File: internal/provider/provider.go (hunk 5 above)
Comment (danielfurman) [reply internal/provider/provider.go:180]: To make it clear: we have removed an option to configure retryable status code in Kentik Go client to simplify the code. @jdziurzynski pointed to the code responsible for configuring gRPC retries in the client.

Comment (mmac-m3a) [reply internal/provider/provider.go:180]: OK. We should retry any transient errors, i.e. RESOURCE_EXHAUSTED (8), FAILED_PRECONDITION (9), ABORTED (10), UNAVAILABLE (14).

Comment (danielfurman) [reply internal/provider/provider.go:180]: Ticket created: https://clkentik.atlassian.net/browse/KNTK-404

File: internal/provider/resource_cloudexport.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (danielfurman) [reply internal/provider/resource_cloudexport.go:55]: Done

File: examples/data-sources/kentik-cloudexport_item/run.sh [hunk 24]
```diff
@@ -1,6 +1,5 @@
 #!/usr/bin/env bash
 # The script applies the example Terraform configuration.
-# The provider uses stub Kentik API server by default.
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply examples/data-sources/kentik-cloudexport_item/run.sh:3]: Done

File: internal/provider/cloudexport_schema.go [hunk 25]
```diff
@@ -363,20 +358,35 @@ func resourceDataToCloudExport(d *schema.ResourceData) (*models.CloudExport, err
...
+		return nil, fmt.Errorf("bgp properties error: %s", err)
+	}
+	if m != nil {
+		resourceDataToBGPProperties(&export, m)
+	}
+
+	return &export, nil
+}
+
+func resourceDataToCEProperties(export *models.CloudExport, cloudProvider string, d *schema.ResourceData) error {
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply internal/provider/cloudexport_schema.go]: Done

File: internal/provider/cloudexport_schema.go [hunk 26]
```diff
@@ -363,20 +358,35 @@ func resourceDataToCloudExport(d *schema.ResourceData) (*models.CloudExport, err
...
 
 	// required
 	cloudProvider := d.Get("cloud_provider").(string) //nolint: errcheck, forcetypeassert // type enforced by schema
 	export.CloudProvider = models.CloudProvider(cloudProvider)
 
+	if err := resourceDataToCEProperties(&export, cloudProvider, d); err != nil {
+		return nil, err
+	}
+
+	m, err := getObjectFromNestedResourceData(d.Get("bgp"))
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply internal/provider/cloudexport_schema.go]: Done

File: internal/provider/cloudexport_schema.go (hunk 7 above)
Comment (danielfurman) [reply internal/provider/cloudexport_schema.go]: Done

File: internal/provider/cloudexport_schema.go (hunk 8 above)
Comment (danielfurman) [reply internal/provider/cloudexport_schema.go]: Done

File: internal/provider/resource_cloudexport.go (hunk 9 above)
Comment (danielfurman) [reply internal/provider/resource_cloudexport.go]: Done

**Review by danielfurman — CHANGES_REQUESTED** [review]
Good job. After .idea files are removed, it is good to go.

File: internal/provider/data_source_cloudexport_item_test.go [hunk 27]
```diff
@@ -1,33 +1,35 @@
...
 					resource.TestCheckResourceAttr(ceAWSDS, "enabled", "true"),
 					resource.TestCheckResourceAttr(ceAWSDS, "name", "test_terraform_aws_export"),
 					resource.TestCheckResourceAttr(ceAWSDS, "description", "terraform aws cloud export"),
-					resource.TestCheckResourceAttr(ceAWSDS, "api_root", "http://localhost:8080/api"),
-					resource.TestCheckResourceAttr(ceAWSDS, "flow_dest", "http://localhost:8080/flow"),
 					resource.TestCheckResourceAttr(ceAWSDS, "plan_id", "11467"),
 					resource.TestCheckResourceAttr(ceAWSDS, "cloud_provider", "aws"),
+					resource.TestCheckResourceAttr(ceAWSDS, "bgp.0.apply_bgp", "true"),
+					resource.TestCheckResourceAttr(ceAWSDS, "bgp.0.use_bgp_device_id", "dummy-device-id"),
+					resource.TestCheckResourceAttr(ceAWSDS, "bgp.0.device_bgp_type", "dummy-device-bgp-type"),
```
Comment (mmac-m3a) [inline internal/provider/data_source_cloudexport_item_test.go:32]: Is the indentation on lines 36 - 44 really off, or is it just a GitHub UI artifact?

(Note: the comment must be here, because stupid does not allow to put it on line 36)

**Review by mmac-m3a — APPROVED** [review]
LGTM.

File: internal/provider/data_source_cloudexport_item_test.go (hunk 27 above)
Comment (danielfurman) [reply internal/provider/data_source_cloudexport_item_test.go:32]: It is GH UI artifact. CI requires the code to be gofmted
